scalameta/parsers/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala

```diff
@@ -469,54 +458,25 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
     }
   }
 
-  def onlyAcceptMod[M <: Mod, T <: Token](mods: List[Mod])
-     (implicit validMod: ValidModifier[M, T],
-               classifier: Classifier[Mod, M],
+  def onlyAcceptMod[M <: Mod, T <: Token](mods: List[Mod], errorMsg: String)
+     (implicit classifier: Classifier[Mod, M],
                tag: ClassTag[M],
                tokenInfo: TokenInfo[T]) = {
     if (token.isNot[T]) {
-      val errorMsg = validMod.getErrorMessage
       mods.getAll[M].foreach(m => syntaxError(errorMsg, at = m))
     }
   }
 
-  trait ValidModifier[M <: Mod, T <: Token] {
-    def getErrorMessage: String
-  }
-
-  implicit object ValidLazyInVal extends ValidModifier[Mod.Lazy, KwVal] {
-    val getErrorMessage = "lazy not allowed here. Only vals can be lazy"
-  }
-
-  trait InvalidModifier[M <: Mod, T <: Token] {
-    def getErrorMessage: String
-  }
-
-  implicit object InvalidImplicitInTrait extends InvalidModifier[Mod.Implicit, KwTrait] {
-    val getErrorMessage = "traits cannot be implicit"
-  }
-
-  implicit object InvalidImplicitInType extends InvalidModifier[Mod.Implicit, KwType] {
-    val getErrorMessage = "`implicit' modifier can be used only for values, variables, methods and classes"
-  }
-
-  class InvalidSealed[M <: Mod, T <: Token] extends InvalidModifier[M, T] {
-    val getErrorMessage = "`sealed' modifier can be used only for classes"
-  }
-  implicit object InvalidSealedInVal extends InvalidSealed[Mod.Sealed, KwVal]
-  implicit object InvalidSealedInVar extends InvalidSealed[Mod.Sealed, KwVar]
-  implicit object InvalidSealedInDef extends InvalidSealed[Mod.Sealed, KwDef]
-  implicit object InvalidSealedInType extends InvalidSealed[Mod.Sealed, KwType]
-  implicit object InvalidSealedInObj extends InvalidSealed[Mod.Sealed, KwObject]
-
   trait InvalidModCombination[M1 <: Mod, M2 <: Mod] {
-    def getErrorMessage: String
-    def illegalCombination(m1: M1, m2: M2): String =
+    val m1: M1
+    val m2: M2
```

Comment:
It will be clumsy and error-prone, since I would need to parse the runtimeClass string representation. Latest version using params in class definition is much better and clear. I would suggest we keep it as-is there.